x/data/server/query_server.go

```diff
@@ -10,55 +15,106 @@ import (
 var _ data.QueryServer = serverImpl{}
 
 func (s serverImpl) ByHash(ctx types.Context, request *data.QueryByHashRequest) (*data.QueryByHashResponse, error) {
-	return nil, fmt.Errorf("not implemented")
-	//cid := request.Cid
-	//
-	//var timestamp gogotypes.Timestamp
-	//store := ctx.KVStore(s.storeKey)
-	//bz := store.Get(AnchorKey(cid))
-	//if len(bz) == 0 {
-	//	return nil, status.Error(codes.NotFound, "CID not found")
-	//}
-	//err := timestamp.Unmarshal(bz)
-	//if err != nil {
-	//	return nil, err
-	//}
-	//
-	//var signers []string
-	//cidSignerPrefixKey := CIDSignerIndexPrefix(CIDBase64String(cid))
-	//prefixStore := prefix.NewStore(store, cidSignerPrefixKey)
-	//iterator := prefixStore.Iterator(nil, nil)
-	//
-	//for iterator.Valid() {
-	//	signer := string(iterator.Key())
-	//	signers = append(signers, signer)
-	//	iterator.Next()
-	//}
-	//
-	//content := store.Get(DataKey(cid))
-	//
-	//return &data.QueryByCidResponse{
-	//	Timestamp: &timestamp,
-	//	Signers:   signers,
-	//	Content:   content,
-	//}, err
+	iri, err := request.Hash.ToIRI()
+	if err != nil {
+		return nil, err
+	}
+
+	store := ctx.KVStore(s.storeKey)
+	id := s.iriIDTable.GetOrCreateID(store, []byte(iri))
+
+	entry, err := s.getEntry(store, id)
+	if err != nil {
+		return nil, err
+	}
+
+	return &data.QueryByHashResponse{
+		Entry: entry,
+	}, nil
+}
+
+func (s serverImpl) getEntry(store sdk.KVStore, id []byte) (*data.ContentEntry, error) {
+	bz := store.Get(AnchorTimestampKey(id))
+	if len(bz) == 0 {
+		return nil, status.Error(codes.NotFound, "not found")
+	}
+
+	var timestamp gogotypes.Timestamp
+	err := timestamp.Unmarshal(bz)
+	if err != nil {
+		return nil, err
+	}
+
+	var signerEntries []*data.SignerEntry
+	idSignerIndexPrefix := IDSignerIndexPrefix(id)
+	prefixStore := prefix.NewStore(store, idSignerIndexPrefix)
+	iterator := prefixStore.Iterator(nil, nil)
+
+	for iterator.Valid() {
+		signer := sdk.AccAddress(iterator.Key())
+		var timestamp gogotypes.Timestamp
+		err = timestamp.Unmarshal(iterator.Value())
+		if err != nil {
+			return nil, err
+		}
+
+		signerEntries = append(signerEntries, &data.SignerEntry{
```

Comment:
Just thinking about this now – is there an attack vector in someone appending an arbitrary number of signatures to a given ContentHash, and this query getting bloated / slow? especially without pagination of signers...

Reply:
i agree here, i think it would be better to omit signers from the ByHash query and add a new query for a paginated list of signers on a certain content entry